Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,2354 @@
+namespace
+{
+
+// Flip sign of curvature, such that tangentDot = curvature * normal
+Real calcSurfaceCurvature(
+    const ContactGeometry& geometry,
+    const FrenetFrame& X_S,
+    CoordinateAxis direction)
+{
+    return -geometry.calcSurfaceCurvatureInDirection(
+        X_S.p(),
+        getAxis(X_S, direction));
+}
+
+// Helper function for flipping the sign such that positive value is outside of
+// the surface.
+Real calcSurfaceValue(const ContactGeometry& geometry, const Vec3& point_S)
+{
+    return -geometry.calcSurfaceValue(point_S);
+}
+// Flipping sign...
+Vec3 calcSurfaceGradient(const ContactGeometry& geometry, const Vec3& point_S)
+{
+    return -geometry.calcSurfaceGradient(point_S);
+}
+// Flipping sign...
+Mat33 calcSurfaceHessian(const ContactGeometry& geometry, const Vec3& point_S)
+{
+    return -geometry.calcSurfaceHessian(point_S);
+}
+
+} // namespace
+
+//==============================================================================
+//                          CABLE SUBSYSTEM
+//==============================================================================
+
+bool CableSubsystem::isInstanceOf(const Subsystem& s)
+{
+    return Impl::isA(s.getSubsystemGuts());
+}
+
+const CableSubsystem& CableSubsystem::downcast(const Subsystem& s)
+{
+    assert(isInstanceOf(s));
+    return static_cast<const CableSubsystem&>(s);
+}
+
+CableSubsystem& CableSubsystem::updDowncast(Subsystem& s)
+{
+    assert(isInstanceOf(s));
+    return static_cast<CableSubsystem&>(s);
+}
+
+CableSubsystem::CableSubsystem()
+{
+    adoptSubsystemGuts(new Impl());
+}
+
+CableSubsystem::CableSubsystem(MultibodySystem& mbs)
+{
+    adoptSubsystemGuts(new Impl());
+    mbs.adoptSubsystem(*this);
+}
+
+int CableSubsystem::getNumCables() const
+{
+    return getImpl().getNumCables();
+}
+
+const CableSpan& CableSubsystem::getCable(CableSpanIndex ix) const
+{
+    return getImpl().getCable(ix);
+}
+
+CableSpan& CableSubsystem::updCable(CableSpanIndex ix)
+{
+    return updImpl().updCable(ix);
+}
+
+//------------------------------------------------------------------------------
+//                        CABLE SUBSYSTEM IMPL
+//------------------------------------------------------------------------------
+
+const CableSubsystem::Impl& CableSubsystem::getImpl() const
+{
+    return SimTK_DYNAMIC_CAST_DEBUG<const Impl&>(getSubsystemGuts());
+}
+
+CableSubsystem::Impl& CableSubsystem::updImpl()
+{
+    return SimTK_DYNAMIC_CAST_DEBUG<Impl&>(updSubsystemGuts());
+}
+
+int CableSubsystem::Impl::calcDecorativeGeometryAndAppendImpl(
+    const State& state,
+    Stage stage,
+    Array_<DecorativeGeometry>& decorations) const
+{
+    if (stage != Stage::Position)
+        return 0;
+
+    for (const CableSpan& cable : cables) {
+        int returnValue = cable.getImpl().calcDecorativeGeometryAndAppend(
+            state,
+            stage,
+            decorations);
+        if (returnValue != 0) {
+            return returnValue;
```

Comment:
Doesn't this implementation only append the first decorative geometry and then exit? The loop doesn't go over all cables and append all decorations, which is what i'd expect.

Reply:
`SubsystemGuts` has a virtual method `calcDecorativeGeometryAndAppendImpl` which returns an `int`. Looking at `SymbodyMatterSubsystemRep` for example, it does not appear to be a counter, so I assumed it was an error flag.
I see this is also the case for `SubsystemGuts::realizeSubsystemTopologyImpl() -> int` for example. The return value doesn't appear to be handled anywhere though.